### PR TITLE
Csharp Enums

### PR DESCRIPTION
**Pull request by justusth**
I implemented enums for C#, I had some problems with the merge but it looks good now.

I also added so default values are sent, we are sending data to C modules and they require all the fields of the api to be set. We have a pretty large system stack with a transaction system that runs C on VMS, C# modules, nodejs servers to route messages etc and hopefully you can agree that this change is reasonable.

### CODE REVIEW

**Comment by wing328** [comment]
@justusth thanks for the PR. We'll review and let you know if we've any questions.

**Comment by wing328** [comment]
About default values, I would suggest we do the following:

1) keep the current logic `EmitDefaultValue=false` as API clients in other programming languages behave that way (we want to deliver a consistent experience across different languages if possible)
2) ideally instead of hardcoding the behaviour, we want to make it configurable via a property in the Configuration class (e.g. a static property name `PropertyEmitDefaultValue`)

Would using `OnSerializing` callback make EmitDefaultValue configurable?

If that's not possible, we can add another CliOption to determine the value of EmitDefaultValue (true or false) when generating the C# API client.

**Comment by wing328** [comment]
@justusth please do a rebase on the latest master to resolve the merge conflicts.

**Comment by justusth** [comment]
On monday first thing when back from travels. Thank you for helping me with this. What do I need to change EmitDefaultValue=false to? I would prefere making it a configuration issue.

**Comment by wing328** [comment]
@justusth you can refer to this [PR](https://github.com/swagger-api/swagger-codegen/pull/1843) to add a CLI option to specify the default value of `EmitDefaultValue`. If you need any help, please reply to let us know.

Enjoy your travels!

**Comment by ktjn** [comment]
@wing328 @justusth I did a rebase of this branch against master, in git@github.com:ktjn/swagger-codegen.git, https://github.com/ktjn/swagger-codegen

I also made enums nullable as everything else.

Was there anything else needed?

**Comment by wing328** [comment]
@ktjn thanks! Would you please submit a PR and we'll review accordingly?

**Comment by ktjn** [comment]
@wing328 will do after i sync with @justusth

**Comment by justusth** [comment]
We still have to add option to send default values but we should do this directly.

> [PR](https://github.com/swagger-api/swagger-codegen/pull/1843) add a CLI option to specify the default value of EmitDefaultValue. If you need any help, please reply to let us know.

**Comment by wing328** [comment]
@justusth  For CLI option to specify the default value of EmitDefaultValue , I can take are of it as I've added many CLI option before :)

@ktjn please submit a PR when you've time.

**Comment by ktjn** [comment]
@wing328 this pull request is now synced with my branch, but the java 7 build broke for some reason, we will investigate this.

Btw, is this cli option still needed when enums are nullable?

**Comment by wing328** [comment]
Yes, it's still needed.

I'll look into the test and restart it if needed so no worry.

**Comment by ktjn** [comment]
@wing328 ran the tests locally with java 7 without any problems, probably just need to be rerun

**Comment by ktjn** [comment]
@wing328 nice, we will fix the cli option for defaultvalue as well

**Comment by wing328** [comment]
@ktjn do you mind updating test case in the C# Petstore solution (samples/client/petstore/csharp/SwaggerClientTest/SwaggerClientTest.sln ) ?

I got 3 errors related to status. Updating those to enum should resolve the issue. 

Then run the unit tests to ensure all tests pass.

**Comment by ktjn** [comment]
@wing328 i tried a regen and got some other, probable unrelated, error with this:
ApiResponse<ApiResponse> UploadFileWithHttpInfo (long? petId, string additionalMetadata = null, Stream file = null); 

Do you have any clue about it?

**Comment by wing328** [comment]
I didn't get the error you mentioned. Here is the error I got:

screen shot 2016-03-02 at 11 41 23 pm

Did you run the following to regenerate the C# Petstore client?
`mvn clean && ./bin/csharp-petstore.sh` for Mac/Linux
```
mvn clean
.\bin\windows\csharp-petstore.bat
```
for Windows

**Comment by ktjn** [comment]
No, I generated by hand. So, I regenerated the test code and fixed everything. I issued a PR to @justusth, he will update this PR. Thanks for the hint.

**Comment by wing328** [comment]
OK. I'll review later then

**Comment by ktjn** [comment]
@wing328 I added a cli option and did a regen of the test code. Feel free to review :-)

File: samples/client/petstore/java/default/src/main/java/io/swagger/client/model/Pet.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import io.swagger.annotations.ApiModel;
 import io.swagger.annotations.ApiModelProperty;
 import io.swagger.client.model.Category;
+import java.util.*;
```
Comment (wing328) [inline samples/client/petstore/java/default/src/main/java/io/swagger/client/model/Pet.java]: I don't think this change is required for this PR.

We'll update (remove) it in our next Java-related PR.

Comment (ktjn) [inline samples/client/petstore/java/default/src/main/java/io/swagger/client/model/Pet.java]: Should we remove it? I think it comes from a regen of the test code

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java [hunk 2]
```diff
@@ -253,6 +264,17 @@ public void setOptionalAssemblyInfoFlag(boolean flag) {
         this.optionalAssemblyInfoFlag = flag;
     }
 
+    @Override
+    public CodegenModel fromModel(String name, Model model, Map<String, Model> allDefinitions) {
+        CodegenModel codegenModel = super.fromModel(name, model, allDefinitions);
+     if (allDefinitions != null && codegenModel != null && codegenModel.parent != null && codegenModel.hasEnums) {
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/CSharpClientCodegen.java]: For your next PR, would be good if you can ident the code following the Java style guide: https://google.github.io/styleguide/javaguide.html

File: samples/client/petstore/java/default/src/main/java/io/swagger/client/model/Pet.java (hunk 1 above)
Comment (wing328) [inline samples/client/petstore/java/default/src/main/java/io/swagger/client/model/Pet.java]: No worry. Next Java PR should take care of it (as generating the petstore sample should be done before submitting the PR)

**Comment by wing328** [comment]
@ktjn I tested `optionalEmitDefaultValues` (set to true) but didn't work for me. Does it work for you?

**Comment by ktjn** [comment]
@wing328 No :-) I only verified the default behaviour. I will check it. Sorry.

**Comment by wing328** [comment]
@justusth no worry. Please indent line 270 in `CSharpClientCodegen.java` and undo the change in `samples/client/petstore/java/default/src/main/java/io/swagger/client/model/Pet.java` for your next PR.

**Comment by ktjn** [comment]
@wing328 fixed everything now :-)

**Comment by wing328** [comment]
Test results look good. Thanks for the contribution.

**Comment by ktjn** [comment]
@wing328 nice, please let us know if there are anything.

**Comment by justusth** [comment]
@wing328 and thank you, it was fun

**Pull request merged**